[css-text-4] Implement multi-value syntax for white-space CSS property

nt1m opened the pull request:
#### 30f3214ab2d13ea5977cdba2f97ea230a4d5d5d9
<pre>
[css-text-4] Implement multi-value syntax for white-space CSS property
<a href="https://bugs.webkit.org/show_bug.cgi?id=256926">https://bugs.webkit.org/show_bug.cgi?id=256926</a>
rdar://109486180

Reviewed by Antti Koivisto and Darin Adler.

Implement `&lt;white-space-collapse&gt; || &lt;text-wrap&gt;` syntax from: <a href="https://drafts.csswg.org/css-text-4/#white-space-property">https://drafts.csswg.org/css-text-4/#white-space-property</a>

This allows new combinations to be expressed using the shorthand, notably:
- white-space-collapse: preserve-breaks; with text-wrap: nowrap;
- white-space-collapse: break-spaces; with text-wrap: nowrap;

Gate the new syntax behind the CSSWhiteSpaceLonghandsEnabled preference, since we don&apos;t have full layout support for those new combinations yet.

Also remove discard/preserve-spaces values from white-space-collapse that we don&apos;t plan to implement in the layout engine in the near future.
They can be re-added later easily once we get to implementing them.

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand-expected.txt:
Rebaseline now passing test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/white-space-expected.txt:
Rebaseline test now that we removed discard/preserve-spaces.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
Remove white-space-collapse values that aren&apos;t planned in the near future (discard &amp; preserve-spaces).

* Source/WebCore/css/CSSProperties.json:
Remove white-space-collapse values that aren&apos;t planned in the near future (discard &amp; preserve-spaces).
Also export parsers for white-space-collapse/text-wrap so they can be reused in CSSPropertyParser.cpp.

* Source/WebCore/css/CSSValueKeywords.in:
Remove white-space-collapse values that aren&apos;t planned in the near future (discard &amp; preserve-spaces).

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::whiteSpaceShorthandValue):
Implement computed style serialization for the previously unsupported combinations of text-wrap/white-space-collapse.

* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serializeWhiteSpace const):
Implement specified style serialization for the previously unsupported combinations of text-wrap/white-space-collapse.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeWhiteSpaceShorthand):
Add parsing support for the multi-value syntax.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
Remove white-space-collapse values that aren&apos;t planned to be implemented in the layout engine (discard &amp; preserve-spaces).

Canonical link: <a href="https://commits.webkit.org/265588@main">https://commits.webkit.org/265588@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70ee27c1924b020217f4f9d950b52a7a68242c90

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11330 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11541 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11882 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12981 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10803 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11351 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13920 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11525 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13710 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11492 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12368 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/9592 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13398 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9663 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10271 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17462 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10732 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/10428 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/13641 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8922 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10015 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2713 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14291 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10699 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2550,43 +2550,72 @@ bool CSSPropertyParser::consumeListStyleShorthand(bool important)
-    case CSSValueNormal:
-        whiteSpaceCollapse = CSSPrimitiveValue::create(CSSValueCollapse);
-        textWrap = CSSPrimitiveValue::create(CSSValueWrap);
-        break;
-    case CSSValueNowrap:
+    // Fill in default values if one was missing from the multi-value syntax.
+    if (!whiteSpaceCollapse)
         whiteSpaceCollapse = CSSPrimitiveValue::create(CSSValueCollapse);
-        textWrap = CSSPrimitiveValue::create(CSSValueNowrap);
-        break;
-    case CSSValuePre:
-        whiteSpaceCollapse = CSSPrimitiveValue::create(CSSValuePreserve);
-        textWrap = CSSPrimitiveValue::create(CSSValueNowrap);
-        break;
-    case CSSValuePreLine:
-        whiteSpaceCollapse = CSSPrimitiveValue::create(CSSValuePreserveBreaks);
-        textWrap = CSSPrimitiveValue::create(CSSValueWrap);
-        break;
-    case CSSValuePreWrap:
-        whiteSpaceCollapse = CSSPrimitiveValue::create(CSSValuePreserve);
+    if (!textWrap)
         textWrap = CSSPrimitiveValue::create(CSSValueWrap);
```
Maybe writing the whole multi-value syntax in a lambda would help to make the code easier to follow (especially the last lines 2614 - 2618 which looks like general code but only exist for the multi value case) ?

nt1m replied:
Not sure it really improves readability IMO, because it'd just be a lambda that's called once

darinadler commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1171,20 +1171,24 @@ String ShorthandSerializer::serializeWhiteSpace() const
+    if (textWrap == CSSValueWrap)
+        return nameString(whiteSpaceCollapse);
+
+    return makeString(nameString(whiteSpaceCollapse), ' ', nameString(textWrap));
```
These should use `nameLiteral`.

darinadler replied:
Oh sorry, @nt1m, I was way too terse. I meant that two cases here that are arguments to `makeString` should use `nameLiteral` for better efficiency, since appending an `ASCIILiteral` can be more efficient than doing the same with a `String`. All the ones that return a value to a `String` should use `nameString` for better efficiency, using the existing string rather than creating a new one.

darinadler reviewed (approved):
Comments say we don’t plan to implement `discard` and `preserve-spaces` but do not say why.

nt1m committed:
[css-text-4] Implement multi-value syntax for white-space CSS property
https://bugs.webkit.org/show_bug.cgi?id=256926
rdar://109486180

Reviewed by Antti Koivisto and Darin Adler.

Implement `<white-space-collapse> || <text-wrap>` syntax from: https://drafts.csswg.org/css-text-4/#white-space-property

This allows new combinations to be expressed using the shorthand, notably:
- white-space-collapse: preserve-breaks; with text-wrap: nowrap;
- white-space-collapse: break-spaces; with text-wrap: nowrap;

Gate the new syntax behind the CSSWhiteSpaceLonghandsEnabled preference, since we don't have full layout support for those new combinations yet.

Also remove discard/preserve-spaces values from white-space-collapse that we don't plan to implement in the layout engine in the near future.
They can be re-added later easily once we get to implementing them.

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand-expected.txt:
Rebaseline now passing test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/white-space-expected.txt:
Rebaseline test now that we removed discard/preserve-spaces.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
Remove white-space-collapse values that aren't planned in the near future (discard & preserve-spaces).

* Source/WebCore/css/CSSProperties.json:
Remove white-space-collapse values that aren't planned in the near future (discard & preserve-spaces).
Also export parsers for white-space-collapse/text-wrap so they can be reused in CSSPropertyParser.cpp.

* Source/WebCore/css/CSSValueKeywords.in:
Remove white-space-collapse values that aren't planned in the near future (discard & preserve-spaces).

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::whiteSpaceShorthandValue):
Implement computed style serialization for the previously unsupported combinations of text-wrap/white-space-collapse.

* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serializeWhiteSpace const):
Implement specified style serialization for the previously unsupported combinations of text-wrap/white-space-collapse.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeWhiteSpaceShorthand):
Add parsing support for the multi-value syntax.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
Remove white-space-collapse values that aren't planned to be implemented in the layout engine (discard & preserve-spaces).

Canonical link: https://commits.webkit.org/265588@main
